presto-main/src/test/java/com/facebook/presto/sql/planner/TestPredicatePushdown.java

```diff
@@ -60,4 +61,35 @@ public void testPushDownToLhsOfSemiJoin()
                                                         "LINE_QUANTITY", "quantity")))),
                                 anyTree(tableScan("orders", ImmutableMap.of("ORDERS_ORDER_KEY", "orderkey"))))));
     }
+
+    @Test
+    public void testPredicateFromFilterSidePropagatesToSourceSideOfSemiJoin()
+    {
+        assertPlan("SELECT quantity FROM (SELECT * FROM lineitem WHERE orderkey IN (SELECT orderkey FROM orders where orderkey = 2))",
+                anyTree(
+                        semiJoin("LINE_ORDER_KEY", "ORDERS_ORDER_KEY", "SEMI_JOIN_RESULT",
+                                anyTree(
+                                        filter("LINE_ORDER_KEY = BIGINT '2' AND BIGINT '2' = LINE_ORDER_KEY", // TODO this should be simplified
+                                                tableScan("lineitem", ImmutableMap.of(
+                                                        "LINE_ORDER_KEY", "orderkey",
+                                                        "LINE_QUANTITY", "quantity")))),
+                                anyTree(
+                                        filter("ORDERS_ORDER_KEY = BIGINT '2'",
+                                                tableScan("orders", ImmutableMap.of("ORDERS_ORDER_KEY", "orderkey")))))));
+    }
+
+    @Test
+    public void testPredicateFromFilterSideNotPropagatesToSourceSideOfSemiJoinIfNotIn()
+    {
+        assertPlan("SELECT quantity FROM (SELECT * FROM lineitem WHERE orderkey NOT IN (SELECT orderkey FROM orders where orderkey = 2))",
+                anyTree(
+                        semiJoin("LINE_ORDER_KEY", "ORDERS_ORDER_KEY", "SEMI_JOIN_RESULT",
+                                project(
+                                        tableScan("lineitem", ImmutableMap.of(
```

Comment:
There could be a filter above TS, w `ORDERS_ORDER_KEY != 2`, right? Add a comment that it's possible, but not implemented yet.

Reply:
I'm wrong. There can't be, because the filtering side may come empty. Please add:
```
// There should be no Filter above table scan, because we don't know whether SemiJoin's filtering source is empty.
```

Reply:
Extended a bit.